dotnet/CoreLib/Pipeline/DistributedPipelineOrchestrator.cs

```diff
@@ -167,28 +184,30 @@ public override async Task RunPipelineAsync(DataPipeline pipeline, CancellationT
 
     private async Task MoveForwardAsync(DataPipeline pipeline, CancellationToken cancellationToken = default)
     {
-        // Note: the pipeline state is persisted in two places:
-        // * source of truth: in the queue (see the message enqueued)
-        // * async copy: in the container together with files - this can be out of sync and is synchronized on dequeue
-
-        if (pipeline.RemainingSteps.Count == 0)
+        if (pipeline.Complete)
         {
             this.Log.LogInformation("Pipeline '{0}' complete", pipeline.DocumentId);
 
-            // Try to save the pipeline status
-            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken, ignoreExceptions: false).ConfigureAwait(false);
+            // Save the pipeline status. If this fails the system should retry the current step.
+            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
         }
         else
         {
             string nextStepName = pipeline.RemainingSteps.First();
             this.Log.LogInformation("Enqueueing pipeline '{0}' step '{1}'", pipeline.DocumentId, nextStepName);
 
+            // Execute as much logic as possible before writing the new pipeline state to disk,
+            // to reduce the chance of the persisted state to be out of sync.
             using IQueue queue = this._queueClientFactory.Build();
             await queue.ConnectToQueueAsync(nextStepName, QueueOptions.PublishOnly, cancellationToken).ConfigureAwait(false);
-            await queue.EnqueueAsync(ToJson(pipeline), cancellationToken).ConfigureAwait(false);
 
-            // Try to save the pipeline status
-            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken, ignoreExceptions: true).ConfigureAwait(false);
+            // Save the pipeline status to disk.
+            // IMPORTANT: If this fails the system will retry the "next" step stored on disk,
+            await this.UpdatePipelineStatusAsync(pipeline, cancellationToken).ConfigureAwait(false);
```

Comment:
If this fails, wouldn't it throw an uncaught exception, which would cause the whole pipeline to halt? Even if it keeps going since the messages are still there, wouldn't it continue the current step?

Reply:
The pipelines should never halt, thanks to the queues triggering an automatic retry in case of errors.
When we reach this line we just completed the business logic of a specific step, and we're trying to update the pipeline state stored on disk. The queue message that got us here is still in the queue, "locked". If an exception occurs, the exception bubbles up to the queue message dispatcher, where it is caught, the queue message is unlocked, and thus the same message is dispatched again, leading to a retry.